package.json

```diff
@@ -27,7 +27,7 @@
   },
   "scripts": {
     "build": "turbo run build",
-    "dev": "turbo run dev",
+    "dev": "turbo run dev --concurrency=12",
```

Comment:
Does this actually work? I'm having a bit of trouble getting the dev server to open consistently when running `pnpm dev` from the root.

Sometimes it doesn't open at all, while other times it shows an error

![image](https://github.com/iTwin/iTwinUI/assets/9084735/88762c57-d815-4603-a0a5-4778d74fdb3b)

Reply:
That seems to be a different issue. Adding to the concurrency is meant to fix the issue where not all of the dev commands would run concurrently, because there are 11 and the default max is 10. I thought the issue you just posted a screenshot for had been resolved by fixing the pnpm-lock in one of my earlier commits, but I guess not. Let me see what I can figure out.

Reply:
im also working on running tests concurrently. I haven't gotten anything to work yet but was looking into [GNU parallel](https://www.gnu.org/software/parallel/). Would something like this work or are you solving a different issue?

Reply:
What? This has nothing to do with running tests or GNU parallel.

The error is happening because of a race condition between the dev servers of `@itwin/itwinui-react` and `remix-playground`. Remix seems to start up before itwinui-react is ready and doesn't seem to be able to recover later.

Reply:
Yeah, the race condition is hard to solve in this case, since we want both of the dev servers to be running simultaneously. I have found a hacky solution by having the `remix-playground dev` command depend on the `@itwin/itwinui-react build` command, which delays the server startup long enough to prevent the race condition. This is obviously not an ideal solution, since it delays the start of the dev server. I can also create a similar issue with the `@itwin/itwinui-css` and `portal` servers, but the portal is able to recover from the error. It seems that Astro recovers better, but I don't yet know why. Similarly, I can most consistently get these issues to happen when running `dev` immediately after `install`, running `build` first seems to prevent them. Again I don't know why, since the `dev` commands run clean on their startup.

Reply:
This might be an opportunity to improve our `dev` command if you want to look into that.

The goal is simple:
> Running `pnpm dev` should start up the dev servers as fast as possible (ideally under five seconds).

Introducing a dependency on `build` definitely makes it slow, but we don't need everything from `build`. (`tsc` is the slowest part of the process, and we definitely don't need that).

Reply:
I'll take a look. If there is a minimal build that helps, it could fix all issues of this type.

Reply:
Ideally, here's what would happen when running `pnpm dev`:
1. Build `@itwin/itwinui-variables/index.css` and `@itwin/itwinui-css/css/all.css` only if stale version doesn't exist.
   - This should take 0 seconds if file already exists. Maybe a few seconds if fresh build.
2. Transpile all TS(X) files inside `@itwin/itwinui-react` into plain JS.
   - This should take a fraction of a second, because it does not need to run `tsc` to produce `.d.ts` files.
3. Run CSS modules transpilation for producing `styles.css` (see [`styles.js` folder](https://github.com/iTwin/iTwinUI/tree/main/packages/itwinui-react/src/styles.js)).
   - This should take <1 second.
4. Start all dev servers, and also start "watching" SCSS/TS changes inside all `packages`. For every change, basically repeat the above 3 steps on-demand, rather than building the whole thing.
   - This time we will lazily produce `.d.ts` files (note that it was skipped during step 2).

That's pretty close to what already happens today, except steps 1-3 are currently non-blocking, and there are some inconsistencies and imperfections in various places.

---

The "dream" scenario would be to not even need steps 1-3 _at all_. [Turborepo docs](https://turbo.build/repo/docs/handbook/sharing-code/internal-packages) describe it best:

> _Instead of bundling our package ourselves, we're going to make the app which imports the package bundle it for us._

The main hurdle to this right now is that our package.json `exports` explicitly map to `esm`/`cjs` folders, whereas our source code lives in the `src` folder. If we can change the `exports` field only during local development, that might solve all the issues.

https://github.com/iTwin/iTwinUI/blob/ef62ab84978f12540a9d9b3d75606cd6d2039131/packages/itwinui-react/package.json#L10-L15

---

If you're interesting in looking into this, it can be a follow-up task. I'm ok with merging this PR even if the remix `dev` server is in a broken state.

Reply:
Yeah, that definitely sounds interesting. Certainly worth following up on.